File: app/routes/api/drive.js
const router = require('express').Router();
const { drive } = require('../../controllers');
const { verifyAuth } = require('../../controllers/auth.controller')

/**
 * @api {get} /drive/:id GetDrive
 * @apiName GetDrive
 * @apiGroup Drive
 * @apiVersion 0.1.0
 * @apiPermission vary
 * 
 * @apiDescription Get a Drive object. 
 * If user is not a participant, None
 * 
 * @apiParam {String} id drive id
 *
 * @apiSuccess {String} id drive id
 * @apiSuccess {String} dest drive destination
 * @apiSuccess {Date} date drive date
 * @apiSuccess {String} driver drive driver username
 * @apiSuccess {String[]} participants drive participants usernames array
 * 
 * @apiError DriveNotFound Drive Not Found
 * @apiError NotAuthorized Not authorized to get drive
 * @apiError AuthRequired Authentication header not found
 */
router.get('/:id', verifyAuth, drive.getDrive)

/**
 * @api {put} /drive/ AddDrive
 * @apiVersion 0.1.0
 * @apiName AddDrive
 * @apiGroup Drive
 * @apiPermission vary
 * 
 * @apiDescription Add a Drive.
 * Adding user must be a participant of the drive.
 * 
 * @apiParam {String} dest drive destination
 * @apiParam {Date} date drive date
 * @apiParam {String} driver drive driver username
 * @apiParam {String[]} participants drive participants usernames array
 * 
 * @apiSuccess {Boolean} success success boolean
 * 
 * @apiError NotAuthorized Not authorized to add drive, not a participant
 * @apiError AuthRequired Authentication header not found
 */
router.put('/', verifyAuth, drive.addDrive)

/**
 * @api {patch} /drive/:id UpdateDrive
 * @apiVersion 0.1.0
 * @apiName UpdateDrive
 * @apiGroup Drive
 * @apiPermission participants of the drive
 * 
 * @apiDescription Update a Drive.
 * Editing user must be a participant of the drive.
 * 
 * @apiParam {String} id drive id (URL param)
 * @apiParam {String} dest drive destination
 * @apiParam {Date} date drive date
 * @apiParam {String} driver drive driver username
 * @apiParam {String[]} participants drive participants usernames array
 * 
 * @apiSuccess {Boolean} success success boolean
 * 
 * @apiError DriveNotFound Drive Not Found
 * @apiError NotAuthorized Not authorized to update drive, not a participant
 * @apiError AuthRequired Authentication header not found
 */
router.patch('/:id', verifyAuth, drive.updateDrive)

/**
 * @api {delete} /drive/:id DeleteDrive
 * @apiVersion 0.1.0
 * @apiName DeleteDrive
 * @apiGroup Drive
 * @apiPermission participants of the drive
 * 
 * @apiDescription Delete a Drive.
 * Deleting user must be a participant of the drive.
 * 
 * @apiParam {String} id drive id
 * 
 * @apiSuccess {Boolean} success success boolean
 * 
 * @apiError DriveNotFound Drive Not Found
 * @apiError NotAuthorized Not authorized to update drive, not a participant
 * @apiError AuthRequired Authentication header not found
 */
router.delete('/:id', verifyAuth, drive.deleteDrive)

module.exports = router;

File: app/APIClient/drive.py
import helper

def getDrive(driveId):
    resp = helper.request(method="GET", uri="/drive/{}".format(driveId))
    if resp.ok:
        return resp.json()
    return False

def addDrive(dest, date, driver, participants):
    resp = helper.request(method="PUT", uri="/drive",
    body = {
        "dest": dest,
        "date": date,
        "driver": driver,
        "participants": participants
    })
    if resp.ok:
        return resp.json().get('id')
    return False

def deleteDrive(driveId):
    resp = helper.request(method="DELETE", uri="/drive/{}".format(driveId))
    if resp.ok:
        return True
    return False

def updateDrive(driveId, dest=None, date=None, driver=None, participants=None):
    d = {}
    if dest:
        d['dest'] = dest
    if date:
        d['date'] = date
    if driver:
        d['driver'] = driver
    if participants:
        d['participants'] = participants
    resp = helper.request(method="PATCH", uri="/drive/{}".format(driveId),
        body = d)
    if resp.ok:
        return True
    return False